src/main/java/uk/gov/hmcts/reform/sandl.snlevents/service/ProblemService.java

```diff
@@ -3,18 +3,109 @@
 
-    public List<Problem> getProblems() {
-        return problemRepository.findAll();
+    private final Function<ProblemReference, ProblemReferenceResponse> problemReferenceDbToResponse =
+        (ProblemReference pr) -> {
+            ProblemReferenceResponse response = new ProblemReferenceResponse();
+            response.setId(pr.getId());
+            response.setEntity(pr.getEntity());
+            response.setEntityId(pr.getEntityId());
+            response.setDescription(pr.getDescription());
+            response.setProblemId(pr.getProblem().getId());
+            return response;
+        };
+
+    public final Function<Problem, ProblemResponse> problemDbToResponse = (Problem p) -> {
+        ProblemResponse response = new ProblemResponse();
+        response.setId(p.getId());
+        response.setType(p.getType());
+        response.setSeverity(p.getSeverity());
+        response.setReferences(
+            p.getReferences()
+                .stream()
+                .map(problemReferenceDbToResponse)
+                .collect(Collectors.<ProblemReferenceResponse>toList())
+        );
+        return response;
+    };
+
+    private final Function<CreateProblemReference, ProblemReference> problemReferenceCreateToDb =
+        (CreateProblemReference cpr) -> {
+            ProblemReference transformed = new ProblemReference();
+            transformed.setId(UUID.randomUUID().toString());
+            transformed.setEntity(FactTransformer.transformToEntityName(cpr.getFact()));
+            transformed.setEntityId(cpr.getFactId());
+            transformed.setDescription(cpr.getDescription());
+            return transformed;
+        };
+
+    public final Function<CreateProblem, Problem> problemCreateToDb = (CreateProblem cp) -> {
+        Problem transformed = new Problem();
+        transformed.setId(cp.getId());
+        transformed.setType(cp.getType());
+        transformed.setSeverity(cp.getSeverity());
+        transformed.setMessage(cp.getMessage());
+        transformed.setReferences(
+            cp.getReferences()
+                .stream()
+                .map(problemReferenceCreateToDb)
+                .collect(Collectors.<ProblemReference>toList())
+        );
+        transformed.getReferences().forEach(pr -> pr.setProblem(transformed));
+        return transformed;
+    };
+
+    public List<ProblemResponse> getProblems() {
+        final List<Problem> problems = problemRepository.findAll();
+
+        return problems.stream()
+            .map(problemDbToResponse)
+            .collect(Collectors.toList());
+    }
+
+    public Problem save(Problem problem) {
+        return problemRepository.save(problem);
+    }
+
+    public void removeIfExist(String id) {
+        if (problemRepository.exists(id)) {
+            problemRepository.delete(id);
+        }
+    }
+
+    public List<ProblemResponse> getProblemsByReferenceTypeId(String referenceEntityId) {
+        List<Problem> problems = entityManager
```

Comment:
I thought that you could move the query to repo (using @Query annotation) and simply call problemRepository.findProblemsByReference(referenceEntityId) instead of using entityManager here

Reply:
didn't thought about it, will check how